src/validate.js

```diff
@@ -0,0 +1,183 @@
+let debug = require('debug')('taskcluster-lib-validate')
+let _ = require('lodash');
+let fs = require('fs');
+let path = require('path');
+let walk = require('walk')
+let yaml = require('js-yaml');
+let urljoin = require('url-join');
+let assert = require('assert');
+let Ajv = require('ajv');
+let AWS = require('aws-sdk');
+let Promise = require('promise');
+
+async function validator(options) {
+
+  let cfg = _.defaults(options, {
+    constants: './schemas/contants.yml',
+    folder: './schemas',
+    publish: process.env.NODE_ENV == 'production',
+    baseUrl: 'http://schema.taskcluster.net/',
+    bucket: 'schemas.taskcluster.net',
+  });
+
+  if (_.isString(cfg.constants)) {
+    let fullpath = path.resolve(cfg.constants);
+    debug('Attempting to set constants by file: %s', fullpath);
+    try {
+      cfg.constants = yaml.safeLoad(fs.readFileSync(fullpath, 'utf-8'));
+    }
+    catch (err) {
+      if (err.code == 'ENOENT') {
+        debug('Constants file does not exist, setting constants to {}');
+        cfg.constants = {};
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  let promises = [];
+  let schemas = [];
+  let ajv = Ajv({useDefaults: true});
+
+  function addSchema (root, name) {
+    let json = null;
+    let data = fs.readFileSync(path.resolve(root, name), 'utf-8');
+    if (/\.ya?ml$/.test(name)) {
+      json = yaml.safeLoad(data);
+    } else {
+      json = JSON.parse(data);
+    }
+
+    let schema = render(json, cfg.constants);
+
+    name = name.replace(/\.ya?ml$/, '.json');
+    let id = urljoin(cfg.baseurl, cfg.prefix, name) + '#';
+    if (!schema.id) {
+      schema.id = id;
+    }
+    if (schema.id !== id) {
+      debug("Bad schema name: %s expected: %s", schema.id, id);
+      throw new Error('Incorrect schemaId specified. It is recommended not to set' +
+          'one at all. It will be set automatically.');
+    }
+
+    try {
+      ajv.addSchema(schema);
+      debug('Loaded schema with id of "%s"', schema.id);
+      schemas.push({name: name, schema: schema});
+    } catch(err) {
+      debug('failed to load schema at %s', path.resolve(root,name));
+      throw err;
+    }
+  }
+
+  function finishSchemaLoading() {
```

Comment:
Since you are calling `walkSync` you have no need for the `end` listener... So you can just do this after the `let walker = walk.walkSync(path.resolve(cfg.folder), walkOptions);` call... right?